src/app/OperationalDeviceProxy.h

```diff
@@ -72,7 +72,50 @@ struct DeviceProxyInitParams
 
 class OperationalDeviceProxy;
 
-typedef void (*OnDeviceConnected)(void * context, OperationalDeviceProxy * device);
+/**
+ * @brief Helper function as applications refactor away from holding onto a device proxy.
+ *
+ * Deprecated - Avoid using this object.
+ *
+ * This is an object that is meant to only be used by application that previously incorrectly held onto
+ * OperationDeviceProxy in previous implementations of OnDeviceConnected. It is expected that over time
+ * all instances of DeviceProxySession will be refactored out of existence.
+ */
+class DeviceProxySession : public DeviceProxy
+{
+public:
+    DeviceProxySession(Messaging::ExchangeManager * exchangeMgr, const SessionHandle & sessionHandle) :
+        mExchangeMgr(exchangeMgr), mSecureSession(sessionHandle), mPeerScopedNodeId(sessionHandle->GetPeer())
+    {}
+    DeviceProxySession() {}
+
+    void ShutdownSubscriptions() override { VerifyOrDie(false); } // Currently not implemented.
+    void Disconnect() override
+    {
+        // TODO should I be cleaning up mExchangeMgr, and mPeerScopedNodeId.
+        mSecureSession.Release();
+    }
+    Messaging::ExchangeManager * GetExchangeManager() const override { return mExchangeMgr; }
+    chip::Optional<SessionHandle> GetSecureSession() const override { return mSecureSession.Get(); }
+    NodeId GetDeviceId() const override { return mPeerScopedNodeId.GetNodeId(); }
+    ScopedNodeId GetPeerScopedNodeId() const { return mPeerScopedNodeId; }
+
+private:
+    bool IsSecureConnected() const override { return static_cast<bool>(mSecureSession); }
+
+    Messaging::ExchangeManager * mExchangeMgr = nullptr;
```

Comment:
If `OnDeviceConnected` is to deliver a reference (see my previous comment), then this should be too.

Reply:
I am still unsure about making this a reference in `DeviceProxySession`, but could be convinced once I am done this PR and know how `DeviceProxySession` is used everywhere. I need to figure out what happens when `DeviceProxySession` is not initialized at all like on the python side.

If the application allocated this on the heap and frees it when it is done I agree. But if it becomes a global in some places where there is only once instance that is needed and we are no allowed to allocate it on the heap then we might need the ability to not have an instance of ExchangeManager.